docs/layout-of-source-files.rst

```diff
@@ -216,101 +216,562 @@ the code below creates new global symbols ``alias`` and ``symbol2`` which refere
+
+There are several ways to load modules into the virtual filesystem:
+
+#. **import statement**
+
+   The ``import`` statement tells the compiler to load and compile a module and then make its
+   content available to the current module.
+
+   We will refer to the path used in the statement as *import path*.
+   The import path is translated into an import key first and then the compiler uses the key to
+   look up the file in its virtual filesystem.
+
+   Imports can be broadly classified into three categories based on how the path is specified:
+
+   .. code-block:: solidity
+
+       import "/contracts/lib/token.sol";   // Absolute
+       import "contracts/lib/token.sol";    // Relative to base
+       import "./contracts/lib/token.sol";  // Relative to source
+
+   There is actually no distinction between absolute imports and imports relative to base at the
+   virtual filesystem level - in both cases the import path is translated into an import key using
+   the same rules - but they are handled differently by the default file loader.
+
+   Imports relative to source, on the other hand, need to be interpreted as paths and normalized to
+   be properly resolved into import keys.
+   The path in this case must conform to UNIX path conventions regardless of the underlying platform.
+
+#. **CLI**
+
+   To compile a file using the command-line interface of the compiler you specify one or more paths:
+
+   .. code-block:: bash
+
+       solc contract.sol /usr/local/dapp-bin/token.sol
+
+   These are interpreted as *filesystem paths* and the rules for translating them into import keys
+   are different than for import paths.
+   They follow the shell rules instead: they can only be relatve to
+   the current working directory and they are normalized and converted into canonical form according
+   to platform-specific rules to determine which physical file they refer to.
+   Only if the file actually exists under that path, it is added to the virtual filesystem and the
+   import key is based on the canonical path.
+
+#. **Standard JSON (as content)**
+
+   An alternative way to compile your project is to use the ``--standard-json`` option and provide
+   a JSON file containing all of your source code:
+
+   .. code-block:: json
+
+       {
+           "language": "Solidity",
+           "sources": {
+               "contract.sol": {
+                   "content": "import \"./util.sol\";\ncontract C {}"
+               },
+               "util.sol": {
+                   "content": "library Util {}"
+               },
+               "/usr/local/dapp-bin/token.sol": {
+                   "content": "contract Token {}"
+               }
+           },
+           "settings": {"outputSelection": {"*": { "*": ["metadata", "evm.bytecode"]}}}
+       }
+
+   The ``sources`` dictionary specifies the initial content of the virtual filesystem and you
+   can use import keys directly there.
```

Comment:
I'm not sure we should really use term "import key" here. These values are the names of the source  files as the compiler calls them. The import statement is just one way to reference to them.

Reply:
"import key" is exactly the term I am using for "names of the source files as the compiler calls them" :) And is different from "import path" (which is the exact text you put inside `import "...";`).

Is there another name that would be clearer? Maybe "source unit ID"? Or "VFS key"?

Reply:
I have renamed "import key" to "source unit ID" everywhere in the text.